cli/demo: simplify the TLS configuration

knz opened the pull request:
Fixes #94310. 
Fixes #94309. 
Fixes #92541.
Epic: CRDB-22814

TLDR: this patch ensures a separate SQL client can resume the
connection to `cockroach demo` when it is restarted, without
reconfiguring the connection URL. It also enables running `cockroach
sql`against a`demo` session with fewer/simpler arguments.

NB: This does not fix _all_ our woes from CRDB-22814; there will be other PRs to address the remaining issues.

**Change 1 (enhancement): reuse directory**

Prior to this patch, `cockroach demo` was auto-generating new TLS
keys/certs on every run, and using a different directory every time.

Especially the different directory made the tool difficult to use: it
was forcing the connection URLs to change on every invocation.

To fix this, this patch now reuses the same directory,
`$HOME/.cockroach-demo`, for every invocation. Certificate/key files
in that directory are reused if already present.

This also makes it possible for SQL clients to simply reconnect
(reusing the same connection string) when the `demo` session is
restarted.

NB: The reason why this change does not use `$HOME/.cockroach-certs`
is that it should remain possible to run `cockroach demo` alongside a
regular server, and the node certs for a regular server should contain
different metadata than for a demo server (esp. relating to server
addresses).

**Change 2 (enhancement): also produce a TLS client cert for the `demo` user**

Prior to this patch, `demo` would only write a TLS client cert for the
`root` user in the certificate directory. This was preventing direct
use of TLS client cert authentication for the `demo` user,
even by those clients that support TLS client cert authn (such as
`cockroach sql`).

This patch fixes this by also producing the TLS client cert for the
`demo` user.

As a result, running `cockroach sql` against the `demo` cluster
is now possible **without entering a password**.

Passwordless access is also possible from other SQL clients by
providing the `demo` client cert and key in the connection URL.
This is now simplified thanks to change 1 above, as the certificate
directory is now always the same.

**Change 3 (enhancement): show how to use use `cockroach sql`**

Prior to this patch, `demo` would only print out connection URLs on
startup and `\demo ls`. Because the connection URLs are relatively
long, it was making it complex and error-prone to copy-paste them
properly for use in a separate `cockroach sql` session.

This patch prints out an example `cockroach sql` invocation alongside
the connection URLs.

For example:
```
...
(cli) cockroach sql --certs-dir=/data/home/kena/.cockroach-demo -u demo -p 26258
(sql) postgresql://demo:demo99186@127.0.0.1:26258/defaultdb?sslmode=require&sslrootcert=%2Fdata%2Fhome%2Fkena%2F.cockroach-demo%2Fca.crt
...
```

**Change 4 (enhancement): simpler multi-tenant TLS**

Prior to this patch, `cockroach demo` was generating the tenant TLS
configuration using a separate CA key/cert pair. This is not needed
for a demo session, and was also forcing every SQL client that wants
to connect to a secondary tenant server to specify the tenant CA cert
in connection URLs.

This patch simplifies this by reusing the same CA for simulating
secondary tenants. This also enables running `cockroach sql` without
extra arguments to connect to a secondary tenant server.

**Change 5 (enhancement): enable overriding the password using an env var**

As requested in #92541. Although the need for this is now limited
because automated tests can now rely on stable client certs
in `~/.cockroach-demo`.

----

Release note (cli change): The TLS parameters to connect to a
`cockroach demo` cluster from SQL clients has been simplified.

Release note (cli change): The password assigned to the `demo` user in
`cockroach demo` can now be overridden by the env var
`COCKROACH_DEMO_PASSWORD`. This is meant for use in automated tests,
when tests cannot be configured to use TLS client cert authentication
using the client certs in `$HOME/.cockroach-demo`.

knz committed:
cli/demo: simplify the TLS configuration

TLDR: this patch ensures a separate SQL client can resume the
connection to `cockroach demo` when it is restarted, without
reconfiguring the connection URL. It also enables running `cockroach
sql` against a `demo` session with fewer/simpler arguments.

**Change 1 (enhancement): reuse directory**

Prior to this patch, `cockroach demo` was auto-generating new TLS
keys/certs on every run, and using a different directory every time.

Especially the different directory made the tool difficult to use: it
was forcing the connection URLs to change on every invocation.

To fix this, this patch now reuses the same directory,
`$HOME/.cockroach-demo`, for every invocation. Certificate/key files
in that directory are reused if already present.

This also makes it possible for SQL clients to simply reconnect
(reusing the same connection string) when the `demo` session is
restarted.

NB: The reason why this change does not use `$HOME/.cockroach-certs`
is that it should remain possible to run `cockroach demo` alongside a
regular server, and the node certs for a regular server should contain
different metadata than for a demo server (esp. relating to server
addresses).

**Change 2 (enhancement): also produce a TLS client cert for the `demo` user**

Prior to this patch, `demo` would only write a TLS client cert for the
`root` user in the certificate directory. This was preventing direct
use of TLS client cert authentication for the `demo` user,
even by those clients that support TLS client cert authn (such as
`cockroach sql`).

This patch fixes this by also producing the TLS client cert for the
`demo` user.

As a result, running `cockroach sql` against the `demo` cluster
is now possible **without entering a password**.

Passwordless access is also possible from other SQL clients by
providing the `demo` client cert and key in the connection URL.
This is now simplified thanks to change 1 above, as the certificate
directory is now always the same.

**Change 3 (enhancement): show how to use use `cockroach sql`**

Prior to this patch, `demo` would only print out connection URLs on
startup and `\demo ls`. Because the connection URLs are relatively
long, it was making it complex and error-prone to copy-paste them
properly for use in a separate `cockroach sql` session.

This patch prints out an example `cockroach sql` invocation alongside
the connection URLs.

For example:
```
...
(cli) cockroach sql --certs-dir=/data/home/kena/.cockroach-demo -u demo -p 26258
(sql) postgresql://demo:demo99186@127.0.0.1:26258/defaultdb?sslmode=require&sslrootcert=%2Fdata%2Fhome%2Fkena%2F.cockroach-demo%2Fca.crt
...
```

**Change 4 (enhancement): simpler multi-tenant TLS**

Prior to this patch, `cockroach demo` was generating the tenant TLS
configuration using a separate CA key/cert pair. This is not needed
for a demo session, and was also forcing every SQL client that wants
to connect to a secondary tenant server to specify the tenant CA cert
in connection URLs.

This patch simplifies this by reusing the same CA for simulating
secondary tenants. This also enables running `cockroach sql` without
extra arguments to connect to a secondary tenant server.

Release note (cli change): The TLS parameters to connect to a
`cockroach demo` cluster from SQL clients has been simplified.

knz committed:
cli/demo: clarify an argument purpose

Release note: None

knz committed:
cli/demo: don't misuse the `nodeID` type

Prior to this patch, the cluster initialization code for `cockroach
demo` was mistakenly using the type `roachpb.NodeID` to number the
various in-memory servers. This was misguided, because there is no
guarantee that the KV node IDs are allocated using the same sequence
as the server indexes in the `demo` code.

To clarify, this patch changes the logic to properly identify
servers by index and not "node ID".

Release note: None

knz committed:
cli/demo: sanitize the printing of connection URLs

Prior to this patch, the printing of connection URLs for secondary
tenant servers was too simplistic.

Acknowledging that tenant servers are just regular servers,
this commit uses the same code for both types of tenants.

Before:
```
   system tenant
     (webui)    http://127.0.0.1:8083/demologin?password=demo8336&username=demo
     (sql)      postgresql://demo:demo8336@127.0.0.1:26260/defaultdb?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1374536428%2Fca.crt
     (sql/jdbc) jdbc:postgresql://127.0.0.1:26260/defaultdb?password=demo8336&sslmode=require&sslrootcert=%2Ftmp%2Fdemo1374536428%2Fca.crt&user=demo
     (sql/unix) postgresql://demo:demo8336@/defaultdb?host=%2Ftmp%2Fdemo1374536428&port=26260

   tenant 1:
      (sql):  postgresql://demo:demo8336@127.0.0.1:26257/movr?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1374536428%2Fca-client-tenant.crt

   tenant 2:
      (sql):  postgresql://demo:demo8336@127.0.0.1:26258/movr?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1374536428%2Fca-client-tenant.crt
...
```

After:
```
   system tenant
     (webui)    http://127.0.0.1:8083/demologin?password=demo36514&username=demo
     (sql)      postgresql://demo:demo36514@127.0.0.1:26260/defaultdb?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca.crt
     (sql/jdbc) jdbc:postgresql://127.0.0.1:26260/defaultdb?password=demo36514&sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca.crt&user=demo
     (sql/unix) postgresql://demo:demo36514@/defaultdb?host=%2Ftmp%2Fdemo1584871889&port=26260

   tenant 1:
     (webui)    https://127.0.0.1:8080/demologin?password=demo36514&username=demo
     (sql)      postgresql://demo:demo36514@127.0.0.1:26257/movr?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca-client-tenant.crt
     (sql/jdbc) jdbc:postgresql://127.0.0.1:26257/movr?password=demo36514&sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca-client-tenant.crt&user=demo

   tenant 2:
     (webui)    https://127.0.0.1:8081/demologin?password=demo36514&username=demo
     (sql)      postgresql://demo:demo36514@127.0.0.1:26258/movr?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca-client-tenant.crt
     (sql/jdbc) jdbc:postgresql://127.0.0.1:26258/movr?password=demo36514&sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca-client-tenant.crt&user=demo
...
```

Release note: None

knz committed:
cli/demo: sanitize the creation of secondary tenants

Prior to this patch, `cockroach demo --multitenant=true --nodes N` was
creating N different secondary tenants, with one SQL server per
tenant.

This was totally besides the point -- to show multi-node scalability,
we want N different SQL servers _for the same tenant_.

This commit fixes that.
Before:
```
       system tenant
         (webui)    http://127.0.0.1:8083/demologin?password=demo36514&username=demo
         (sql)      postgresql://demo:demo36514@127.0.0.1:26260/defaultdb?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca.crt
         (sql/jdbc) jdbc:postgresql://127.0.0.1:26260/defaultdb?password=demo36514&sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca.crt&user=demo
         (sql/unix) postgresql://demo:demo36514@/defaultdb?host=%2Ftmp%2Fdemo1584871889&port=26260

       tenant 1:
         (webui)    https://127.0.0.1:8080/demologin?password=demo36514&username=demo
         (sql)      postgresql://demo:demo36514@127.0.0.1:26257/movr?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca-client-tenant.crt
         (sql/jdbc) jdbc:postgresql://127.0.0.1:26257/movr?password=demo36514&sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca-client-tenant.crt&user=demo

       tenant 2:
         (webui)    https://127.0.0.1:8081/demologin?password=demo36514&username=demo
         (sql)      postgresql://demo:demo36514@127.0.0.1:26258/movr?sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca-client-tenant.crt
         (sql/jdbc) jdbc:postgresql://127.0.0.1:26258/movr?password=demo36514&sslmode=require&sslrootcert=%2Ftmp%2Fdemo1584871889%2Fca-client-tenant.crt&user=demo

       tenant 3:
...
```

After:
```
     Application tenant:
      (webui)    https://127.0.0.1:8080/demologin?password=demo41322&username=demo
      (sql)      postgresql://demo:demo41322@127.0.0.1:26257/movr?sslmode=require&sslrootcert=%2Ftmp%2Fdemo3364057791%2Fca-client-tenant.crt
      (sql/jdbc) jdbc:postgresql://127.0.0.1:26257/movr?password=demo41322&sslmode=require&sslrootcert=%2Ftmp%2Fdemo3364057791%2Fca-client-tenant.crt&user=demo

     System tenant:
      (webui)    http://127.0.0.1:8083/demologin?password=demo41322&username=demo
      (sql)      postgresql://demo:demo41322@127.0.0.1:26260/defaultdb?sslmode=require&sslrootcert=%2Ftmp%2Fdemo3364057791%2Fca.crt
      (sql/jdbc) jdbc:postgresql://127.0.0.1:26260/defaultdb?password=demo41322&sslmode=require&sslrootcert=%2Ftmp%2Fdemo3364057791%2Fca.crt&user=demo
      (sql/unix) postgresql://demo:demo41322@/defaultdb?host=%2Ftmp%2Fdemo3364057791&port=26260
```

Note: this output will be further simplified once we use a single SQL
listener for multiple tenants; see #92580; and also when we support
using a shared HTTP listener for the demo servers (i.e. when the work
from PR #91744 is extended to `cockroach demo`).

(no release note since this functionality is not yet exposed to end-users)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94323)
<!-- Reviewable:end -->

knz commented:
@aadityasondhi you will also need this. Otherwise the TLS configuration for debug zip will be too complicated.

knz committed:
cli/demo: fix the web autologin for secondary tenants

Release note: None

knz committed:
cli/demo: stabilize the assignment of TCP ports for the RPC service

Prior to this patch, `cockroach demo` would always assign random ports
for the RPC service, as well as a random IP address with `--insecure`
on linux.

This commit fixes that.

Release note: None

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=94323) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=94323) before we can accept your contribution.<br/>**11** out of **12** committers have signed the CLA.<br/><br/>:white_check_mark: yuzefovich<br/>:white_check_mark: knz<br/>:white_check_mark: tbg<br/>:white_check_mark: msbutler<br/>:white_check_mark: jayshrivastava<br/>:white_check_mark: itsbilal<br/>:white_check_mark: rafiss<br/>:white_check_mark: kvoli<br/>:white_check_mark: msirek<br/>:white_check_mark: erikgrinaker<br/>:white_check_mark: mgartner<br/>:x: craig[bot]<br/><hr/>**craig[bot]** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=94323) it.</sub>

knz committed:
cli/demo: make the demo password configurable

Release note (cli change): The password assigned to the `demo` user in
`cockroach demo` can now be overridden by the env var
`COCKROACH_DEMO_PASSWORD`. This is meant for use in automated tests,
when tests cannot be configured to use TLS client cert authentication
using the client certs in `$HOME/.cockroach-demo`.
